### ISSUE TITLE
Document that sd_journal_wait() and related calls only generate events at most every 250ms, i.e. come with latency.

### ISSUE THREAD

**Issue by NickYang24** [systemd/systemd#17574]
**systemd version the issue has been seen with**
 > 246

**Used distribution**
 > fedora32

**Linux kernel version used** (`uname -a`)
 > Linux localhost.localdomain 5.6.6-300.fc32.aarch64

**Expected behaviour you didn't see**
 > 
```
[root@localhost home]# logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages
3008:2020-11-11T21:47:34.973633+08:00 localhost root: AAABbbbbbb
```

**Unexpected behaviour you saw**
 > 
```
[root@localhost home]# logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages
[root@localhost home]# echo $?
1
```

**Steps to reproduce the problem**
 > 
```
logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages
```
It seems there is a delay in rsyslog, so i do a test.

```
[root@localhost ~]# cat rsyslog_speed.sh
#!/bin/bash
function get_ms()
{
        echo $[$(date +%s%N)/1000000]
}

msg="AAABBBB-$RANDOM";
t2=$(get_ms)
logger "$msg"
time { while true; do
        cat /var/log/messages | grep "$msg";
        [ $? -eq 0 ] && break;
done }
t3=$(get_ms)
dt2=$((t3-t2))
echo "rsyslog elapsed:" $dt2 "ms"
```
The time of rsyslog elapsed is about 300ms with systemd v245(even v243) or later.

```
[root@localhost home]# ./rsyslog_speed.sh
Nov 11 21:53:33 localhost root[32238]: AAABBBB-24795

real    0m0.443s
user    0m0.362s
sys     0m0.396s
rsyslog elapsed: 453 ms
```

But with systemd v219, this problem seems dosen't exist, result is listed as below.
```
[root@localhost ~]# ./rsyslog_speed.sh
2020-11-11T21:53:01.275367+08:00 localhost root: AAABBBB-21886

real    0m0.002s
user    0m0.001s
sys     0m0.002s
rsyslog elapsed: 5 ms
```

Then i tested a lot of things, finaly i found the abi that rsyslog used is sd_journal_wait(), i thing there mayba a problem for waiting. So i did another test, the code is listed as below:
```
[root@localhost ~]# cat test.c
#include <stdio.h>
#include <string.h>
#include <systemd/sd-journal.h>
#include <sys/time.h>
#include <syslog.h>
#include <unistd.h>

#define MAX_TESTS 100
#define POLL_TIMEOUT -1

long get_us()
{
        long t_us;
        struct timeval tv;
        gettimeofday(&tv, NULL);
        t_us = tv.tv_sec * 1000000 + tv.tv_usec;
        return t_us;
}

int main(int argc, char *argv[])
{
        long start_us;
        long end_us;
        long delay_us;
        long sum_us = 0;
        int r;
        int count;
        sd_journal *j;

        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
        if (r < 0) {
                fprintf(stderr, "sd_journal_open() failed: %s\n", strerror(-r));
                return r;
        }

        /* jump to tail  */
        r = sd_journal_seek_tail(j);
        r = sd_journal_previous(j);

        count = 0;
        openlog("DelayTest", LOG_CONS, LOG_USER);
        syslog(LOG_INFO, "testmsg%04d\n", count++);
        start_us = get_us();
        while (count < MAX_TESTS) {
                const void *d;
                size_t l;

                r = sd_journal_next(j);
                if (r < 0) {
                        fprintf(stderr, "sd_journal_next() failed: %s\n", strerror(-r));
                        goto final;
                }

                if (r == 0) {
                        printf("into sd_journal_wait\n");
                        /* wait for new messages */
                        r = sd_journal_wait(j, (uint64_t)POLL_TIMEOUT);
                        if (r < 0) {
                                fprintf(stderr, "Failed to wait for changes: %s\n", strerror(-r));
                                break;
                        }
                        continue;
                }

                /* read all until reaching tail */
                r = sd_journal_get_data(j, "MESSAGE", &d, &l);
                if (r < 0) {
                        fprintf(stderr, "Failed to read message field: %s\n", strerror(-r));
                        continue;
                }
                end_us = get_us();
                delay_us = end_us - start_us;
                sum_us += delay_us;
                printf("delay: %10dus\n", delay_us);
                printf("%.*s\n", (int) l, (const char*) d);

                syslog(LOG_INFO, "testmsg%04d\n", count++);
                start_us = get_us();
        }
        printf("average of %d tests: %.2fus\n ", MAX_TESTS, sum_us * 1.0 / (MAX_TESTS));
        closelog();
final:
        sd_journal_close(j);
        return 0;
}
```

The result showed that the time that sd_journal_wait() maybe needed 500ms to exit, the result is listed as below.
```
[root@localhost home]# ./test
into sd_journal_wait
delay:       1662us
MESSAGE=testmsg0000
into sd_journal_wait
delay:     371276us
MESSAGE=testmsg0001
into sd_journal_wait
delay:     500543us
MESSAGE=testmsg0002
into sd_journal_wait
delay:     499282us
MESSAGE=testmsg0003
into sd_journal_wait
delay:     499919us
MESSAGE=testmsg0004
into sd_journal_wait
delay:     499922us
MESSAGE=testmsg0005
into sd_journal_wait
delay:     499932us
...
MESSAGE=testmsg0096
into sd_journal_wait
delay:     499876us
MESSAGE=testmsg0097
into sd_journal_wait
delay:     499915us
MESSAGE=testmsg0098
average of 100 tests: 488648.24us
``` 

But with systemd v219,  the time is about just 300us, so i think maybe there is a delay in sd_journal_wait() that cause the delay of rsyslog delaying logging into /var/log/messages.
```
[root@localhost home]# ./test
into sd_journal_wait
into sd_journal_wait
delay:       2284us
MESSAGE=testmsg0000
into sd_journal_wait
delay:        265us
MESSAGE=testmsg0001
into sd_journal_wait
delay:        249us
MESSAGE=testmsg0002
into sd_journal_wait
delay:        264us
MESSAGE=testmsg0003
into sd_journal_wait
delay:        224us
MESSAGE=testmsg0004
into sd_journal_wait
delay:        254us
MESSAGE=testmsg0005
into sd_journal_wait
delay:        213us
...
MESSAGE=testmsg0094
into sd_journal_wait
delay:        296us
MESSAGE=testmsg0095
into sd_journal_wait
delay:        217us
MESSAGE=testmsg0096
into sd_journal_wait
delay:        210us
MESSAGE=testmsg0097
into sd_journal_wait
delay:        214us
MESSAGE=testmsg0098
average of 100 tests: 259.06us
```

@poettering @keszybz @yuwata

**Comment by vcaputo** [comment]
https://github.com/systemd/systemd/commit/7a24f3bf2fb181243a1957a0cdd54cd919396793

**Comment by NickYang24** [comment]
> [7a24f3b](https://github.com/systemd/systemd/commit/7a24f3bf2fb181243a1957a0cdd54cd919396793)

I will try to patch this to v219. But from the code that i am doubt about if this will cause delaying in sd_journal_wait used by rsyslog? And the delay is about 500ms, double of POST_CHANGE_TIMER_INTERVAL_USEC.

**Comment by poettering** [comment]
the data is written instantly, but the notifications for it are only generated every 250ms, to optimize things. Hence if you look in the files at arbitrary times you'll see the data already being there, but we won't wake you up more often than 250ms about it.

I don#t think there's anything to fix here, the current behaviour makes sense and system logging is entirely asynchronous anyway. 250m is not a long delay. journald is not supposed to give RT guarantees.

I figure we should document this better in the man page of sd_journal_wait() though...

**Comment by keszybz** [comment]
> `[root@localhost home]# logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages`

This is always racy. You're assuming that both journald and syslogd will be fast enough, but there's no guarantee like this. Even if you insert `sleep 0.25` in there, you would still occasionally lose the race.

**Comment by NickYang24** [comment]
> > `[root@localhost home]# logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages`
> 
> This is always racy. You're assuming that both journald and syslogd will be fast enough, but there's no guarantee like this. Even if you insert `sleep 0.25` in there, syslogd would still occasionally lose the race.

Yes, you are right. But before thecommit

**Comment by NickYang24** [comment]
> > `[root@localhost home]# logger AAABbbbbbb; grep -nr AAABbbbbbb /var/log/messages`
> 
> This is always racy. You're assuming that both journald and syslogd will be fast enough, but there's no guarantee like this. Even if you insert `sleep 0.25` in there, syslogd would still occasionally lose the race.

Yes, you are right. But before the commit

**Comment by keszybz** [comment]
Fixed by 696e5a8a73a4441f5aea4b01715d4863051647d6.
